mantidproject/mantid
# More reliable point generation in CSGObject for simple shapes

This PR improves the `CSGObject::generatePointInObject()` methods by introducing functions to find random points in simple shapes using their known geometry. Currently supported shapes are cuboid, cylinder and sphere.

The original methods picked a point in the bounding box and tested if it was within the actual shape. It was possible that none of the random points hit the shape if the shape filled the volume of the bounding box poorly (thin tilted slab, thin diagonal cylinder...). This in turn prevented algorithms like `MonteCarloAbsorption` from running. Note, that the `CSGObject::generatePointInObject()` method which takes an `activeRegion` as an argument may still fail but the probability should be lower than before since the random point needs to hit only the `activeRegion`, not the shape.

Performance tests showed that the bounding box method was actually faster for simple shapes if they filled the bounding box volume nicely whereas the specialized functions were faster for poorly filling shapes. Therefore, we
* try a few times (set to five for the moment) the bounding box method and if it fails
* pick the point using the specialized simple shape functions

For other shapes, we always fall back to the bounding box method.

**Report to:** [nobody].

**To test:**

Check that the script below runs. You can also try running it on master to see if there are any performance differences.

```python
data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1)
data = ConvertUnits(data, Target='Wavelength')
SetSample(data, Geometry={'Shape': 'FlatPlate', 'Width': 3.0, 'Height': 10.0, 'Thick':0.005,
                  'Center': [0.0,0.0,0.0], 'Angle': 135.},
                Material={'ChemicalFormula': '(Li7)2-C-H4-N-Cl6', 'SampleNumberDensity': 0.07},
         )

abscor = MonteCarloAbsorption(data, Interpolation='CSpline')
```

Fixes #24636.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
